Fix memory leak problem when training on MPS device

Aurxs opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->
Fixes https://github.com/ultralytics/ultralytics/issues/22621
I fix memory leak problem when training on MPS device.

First I add memory reclamation strategy after optimizer.step() and each batch.This may mitigate memory leaks by empty torch cache and recycle useless memory.
Second I noticed that the self.loss_items tensor can be separated from the current computation graph which can reduced memory usage without any effect on backpropagation.

I also tested the repair results through a test:
before the fix：
[before_fix.txt](https://github.com/user-attachments/files/23562442/before_fix.txt)
after the fix:
[after_fix.txt](https://github.com/user-attachments/files/23562444/after_fix.txt)
You can see that the maximum memory usage has been significantly reduced after the fix.

I hope my fixes will make some contribution to the community.

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes a memory leak when training on Apple MPS devices by detaching loss tensors and adding targeted MPS memory cleanup during training.

### 📊 Key Changes
- Detaches `self.loss_items` from the computation graph to avoid unnecessary gradient tracking and associated memory growth.
- Adds MPS-specific post-optimizer-step cleanup using `torch.mps.synchronize()` and `torch.mps.empty_cache()`.
- Introduces adaptive garbage collection intervals based on training epoch to more aggressively clean memory in early epochs.
- Adds batch-level MPS cleanup with adaptive intervals to prevent batch data accumulation during training.

### 🎯 Purpose & Impact
- Resolves memory leak issues encountered when training on MPS devices (Apple Silicon), improving training stability.
- Reduces the likelihood of out-of-memory errors and performance degradation over long training runs on MPS.
- Provides safer, more efficient MPS training behavior without affecting non-MPS devices.

Aurxs committed:
Fix mps devices' memory leak problem

Aurxs committed:
detaching loss tensors from computation graph

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant commented on ultralytics/engine/trainer.py:
```diff
@@ -428,6 +428,7 @@ def _do_train(self):
                     self.loss = loss.sum()
                     if RANK != -1:
                         self.loss *= self.world_size
+                    self.loss_items = self.loss_items.detach()  # detach to prevent memory leaks
```
💡 **MEDIUM**: Detaching `self.loss_items` is a good idea for breaking the graph, but it assumes `self.loss_items` is always a tensor. If this can ever be a list, dict, or `None` (e.g., for some custom trainers or callbacks that override it), this will raise an error. Consider either ensuring `self.loss_items` is always a tensor earlier, or guarding the detach with an `isinstance` check.

UltralyticsAssistant commented on ultralytics/engine/trainer.py:
```diff
@@ -436,6 +437,15 @@ def _do_train(self):
                     self.optimizer_step()
                     last_opt_step = ni
 
+                    # MPS-specific memory cleanup after optimizer step
```
⚠️ **HIGH**: `gc.collect()` is used here but the diff does not show an import for the `gc` module. If `gc` is not already imported at the top of this file, this will raise a `NameError` the first time this branch executes. Please ensure `import gc` exists in the module imports.

UltralyticsAssistant commented on ultralytics/engine/trainer.py:
```diff
@@ -465,6 +475,15 @@ def _do_train(self):
 
                 self.run_callbacks("on_train_batch_end")
 
+                # MPS-specific: Clear batch data to prevent accumulation
```
📝 **LOW**: The cleanup logic for MPS (epoch-based `gc_interval` at lines 445–447 and batch-based `batch_interval` at lines 480–483) uses hard-coded thresholds and is duplicated conceptually. To improve clarity and maintainability, consider extracting these interval calculations into a small helper function or configuration option so behavior is easier to tune and keep consistent.

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Overall this PR addresses an important MPS memory issue and introduces targeted cleanup logic, but there are a few concerns: ensure `gc` is imported to avoid runtime errors, verify that `self.loss_items` is always a tensor before calling `.detach()`, and be cautious about the performance cost and hard-coded nature of frequent MPS synchronization and garbage collection inside the training loop. Making the cleanup behavior configurable and centralizing the interval logic would improve maintainability and flexibility.

💬 Posted 3 inline comments

Aurxs committed:
The splitting operation will only be performed if self.loss_items is indeed a tensor.

Aurxs committed:
Merge remote-tracking branch 'origin/fix-mps-memory-leak' into fix-mps-memory-leak

# Conflicts:
#	ultralytics/engine/trainer.py

Aurxs commented:
> ## [Codecov](https://app.codecov.io/gh/ultralytics/ultralytics/pull/22695?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ultralytics) Report
> ❌ Patch coverage is `28.57143%` with `10 lines` in your changes missing coverage. Please review.
> 
> [Files with missing lines](https://app.codecov.io/gh/ultralytics/ultralytics/pull/22695?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ultralytics)	Patch %	Lines
> [ultralytics/engine/trainer.py](https://app.codecov.io/gh/ultralytics/ultralytics/pull/22695?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ultralytics#diff-dWx0cmFseXRpY3MvZW5naW5lL3RyYWluZXIucHk=)	28.57%	[10 Missing ⚠️ ](https://app.codecov.io/gh/ultralytics/ultralytics/pull/22695?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ultralytics)
> 📢 Thoughts on this report? [Let us know!](https://github.com/codecov/feedback/issues/255)

I don't think low Patch coverage is a problem.Because the patch only works on MPS device.It won't run on other divices.

UltralyticsAssistant committed:
Merge branch 'main' into fix-mps-memory-leak

glenn-jocher commented:
Thanks for clarifying; while low patch coverage is understandable for MPS-only paths, we still prefer a small guarded test (for example using `pytest.mark.skipif(not torch.backends.mps.is_available(), ...)`) to touch the new logic so coverage reflects it, which also aligns with our testing guidelines in the [CI/Testing Workflow](https://handbook.ultralytics.com/workflows/ci-testing/).

Aurxs committed:
Add test for MPS memory cleanup after training

Aurxs committed:
Merge branch 'main' into fix-mps-memory-leak

Y-T-G committed:
Remove skipped test for MPS memory cleanup

Removed the test for MPS memory cleanup due to unavailability.

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

Y-T-G committed:
Refactor memory management for MPS compatibility

Removed MPS-specific memory cleanup code and adjusted memory clearing logic.

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

Y-T-G committed:
Remove MPS-specific cleanup in training loop

Removed MPS-specific batch data cleanup to simplify code.

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Y-T-G committed:
Fix import statement for os module

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

Y-T-G committed:
Merge branch 'main' into fix-mps-memory-leak

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good, minimal change to treat MPS differently for memory clearing. One potential blocker: `None` as an argument to `_clear_memory()` may cause a runtime error if the method expects a float threshold and performs numeric comparisons. Recommend ensuring `_clear_memory()` handles `None` safely or switching to an explicit numeric/branch-based approach.

💬 Posted 1 inline comment

Y-T-G committed:
Refactor memory clearing in trainer.py

Updated memory clearing logic to always clear on MPS due to a memory leak issue.

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

Aurxs commented:
@Y-T-G Hello! I have tested the new fix you committed, and here's the result:
[after_fix.txt](https://github.com/user-attachments/files/24187940/after_fix.txt)
As you see, the problem might not be solved. Maybe there're still some problems with it.

By the way, I met an error during my first test in epoch 9:
[error.txt](https://github.com/user-attachments/files/24188035/error.txt)
But the error didn't occur again when I rerun the test. Maybe the fix will call some stability issue?

Y-T-G committed:
Update memory clearing logic for MPS device

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

Aurxs commented:
@Y-T-G I think it works. Here's the test result:
[after_fix.txt](https://github.com/user-attachments/files/24190518/after_fix.txt)
